### PR TITLE
deterministic RR selection

### PR DESCRIPTION
**Pull request by smz115**
Based on: https://github.com/mhmxs/calico-route-reflector-operator/pull/2

### CODE REVIEW

File: bgppeer/bgppeer.go [hunk 1]
```diff
@@ -32,11 +34,14 @@ func (b *BGPPeer) ListBGPPeers() (*calicoApi.BGPPeerList, error) {
 }
 
 func (b *BGPPeer) SaveBGPPeer(peer *calicoApi.BGPPeer) error {
-	if peer.GetUID() == "" {
+	exists, _ := b.CalicoClient.BGPPeers().Get(context.Background(), peer.Name, options.GetOptions{})
```
Comment (mhmxs) [inline bgppeer/bgppeer.go]: This step is wasting resources, would be nice to avoid fetching configs one by one.

File: topologies/multi.go [hunk 2]
```diff
@@ -80,75 +94,85 @@ func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes ma
...
-				if nodeZone == rrZone {
-					if _, ok := rrIndexPerZone[nodeZone]; !ok {
-						rrIndexPerZone[nodeZone] = -1
-					}
-
-					rrsSameZone = append(rrsSameZone, routeReflectors[i])
+		routeReflectorsForNode := t.GetRRsofNode(nodes, existingPeers, n)
+		log.Debugf("Found %v RRs for Node:%s", len(routeReflectorsForNode), n.Name)
+
+		// Select the first two RRs from different zones in Multi Zone clusters
```
Comment (mhmxs) [inline topologies/multi.go:119]: The original implementation also selected RRs at least 2 zones.

File: topologies/multi.go [hunk 3]
```diff
@@ -80,75 +94,85 @@ func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes ma
...
+		// These aren't necessarily from the same cluster as the Node
+		if isMultiZone {
+
+			var firstRR *corev1.Node
+			if len(routeReflectorsForNode) == 0 {
+				firstRR = &routeReflectors[rand.Intn(len(routeReflectors))]
```
Comment (mhmxs) [inline topologies/multi.go:131]: This isn't deterministic :(

File: topologies/multi.go [hunk 4]
```diff
@@ -65,11 +76,14 @@ func (t *MultiTopology) CalculateExpectedNumber(readyNodes int) int {
 	return t.single.CalculateExpectedNumber(readyNodes)
 }
 
-func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes map[*corev1.Node]bool, existingPeers *calicoApi.BGPPeerList) []calicoApi.BGPPeer {
-	bgpPeerConfigs := []calicoApi.BGPPeer{}
+func (t *MultiTopology) GenerateBGPPeers(routeReflectors []corev1.Node, nodes map[*corev1.Node]bool, existingPeers *calicoApi.BGPPeerList) (toRefresh []calicoApi.BGPPeer, toDelete []calicoApi.BGPPeer) {
+	toKeep := map[string]bool{}
 
 	rrConfig := findBGPPeer(existingPeers.Items, DefaultRouteReflectorMeshName)
+	selector := fmt.Sprintf("has(%s)", t.NodeLabelKey)
```
Comment (mhmxs) [inline topologies/multi.go]: Early init of selector, only used later

File: topologies/multi.go (hunk 3 above)
Comment (smz115) [reply topologies/multi.go:131]: It only selects the first RR randomly if there're no RRs configured in BGPPeers for the Node or if none of the configured RRs have the rr-id label [here](https://github.com/mhmxs/calico-route-reflector-operator/pull/3/files#diff-9722ec746f25a527934fd27e1df07de3R67), i.e. the RR status has been removed from the node.

Comment (mhmxs) [reply topologies/multi.go:131]: "It only selects the first RR randomly" so not deterministic at the end of the day :(

File: bgppeer/bgppeer.go (hunk 1 above)
Comment (smz115) [reply bgppeer/bgppeer.go]: Removed, thank you!

File: topologies/multi.go (hunk 4 above)
Comment (smz115) [reply topologies/multi.go]: Moved, thank you!

File: topologies/multi.go (hunk 3 above)
Comment (smz115) [reply topologies/multi.go:131]: At that stage in the RR selection it doesn't have to be deterministic as it's uses the Node's RR state from the BGPPeers configureations.

**Pull request closed without merging**